refactor tunnel middleware into two phases

ekashida opened the pull request:
By splitting the tunnel logic into a detection phase and a handling phase,
applications are free to override the handling logic.

In order to maintain backwards-compatibility, instead of refactoring the
existing tunnel middleware, a tunnel-demux middleware to detect the type of
tunnel request, and tunnel-(rpc|specs|type) middleware to handle each type of
request have been added. Mojito itself will now use these new middleware in
place of the single tunnel middleware, which will remain for those applications
that are specifying it in their configuration.

ekashida committed:
refactor tunnel middleware into two phases

By splitting the tunnel logic into a detection phase and a handling phase,
applications are free to override the handling logic.

In order to maintain backwards-compatibility, instead of refactoring the
existing tunnel middleware, a tunnel-demux middleware to detect the type of
tunnel request, and tunnel-(rpc|specs|type) middleware to handle each type of
request have been added. Mojito itself will now use these new middleware in
place of the single tunnel middleware, which will remain for those applications
that are specifying it in their configuration.

ekashida commented:
Still working on this (tests, docs, end-to-end testing, etc). Would like some feedback for what I have so far!

caridy commented on lib/app/middleware/mojito-handler-tunnel-specs.js:
```diff
@@ -0,0 +1,66 @@
+module.exports = function (config) {
+    return function (req, res, next) {
+        var specsReq = req._tunnel && req._tunnel.specsReq,
+            instance = {},
```
do not create objects if they are not going to be used. move this below the `if (!specsReq)`, otherwise you will be creating an empty object per req even when it is not a tunnel request, in which case this will have to be collected by GC.

ekashida replied:
agreed

caridy commented on lib/app/middleware/mojito-handler-tunnel-specs.js:
```diff
@@ -0,0 +1,66 @@
+
+        type = specsReq.type;
+        name = specsReq.name;
+        name = name && name.split('.').slice(0, -1).join('.');
```
isn't easier to use require('path.extname(), then slicing based on that? this line seems complex.

ekashida replied:
I don't think it was simplified much, but the comment definitely helps :)

```
// Remove the extension name.
ext  = libpath.extname(name);                                                  
name = name && ext && name.slice(0, -ext.length);
```

ekashida replied:
I got rid of this logic completely by passing in (from demux) just the basename without the extension:

```
// Get the basename without the .json extension.                               
name = libpath.basename(path, '.json');
```

caridy commented on lib/app/middleware/mojito-handler-tunnel-demux.js:
```diff
@@ -0,0 +1,120 @@
+        /tunnel/static/{type}/specs/default.json   // according to a UT
+        **/
+
+        url = req.url.split('?')[0];
```
hmm, I guess we should use `require('url').parse()` to really understand that url.

ekashida replied:
agreed

caridy commented on lib/app/middleware/mojito-handler-tunnel-specs.js:
```diff
@@ -0,0 +1,66 @@
+            return sendError(res, 'Not found: ' + req.url, 500);
+        }
+
+        instance.base = type;
```
here is where you use `instance` for the first time, here is where you should create the object with base on it.

ekashida replied:
agreed

caridy commented on lib/app/middleware/mojito-handler-tunnel-specs.js:
```diff
@@ -0,0 +1,66 @@
+        name = name && name.split('.').slice(0, -1).join('.');
+
+        if (!type || !name) {
+            return sendError(res, 'Not found: ' + req.url, 500);
```
If it is not found, it should be 404, if it is an invalid url, then 500 is just fine.

ekashida replied:
Damn you copy/paste!

caridy commented on lib/mojito.js:
```diff
@@ -90,7 +90,10 @@ MojitoServer.MOJITO_MIDDLEWARE = [
     'mojito-parser-body',
     'mojito-parser-cookies',
     'mojito-contextualizer',
-    'mojito-handler-tunnel',
```
this will be problematic for BC. Can we preserve the tunnel as a sugar for all the new middleware? middlewares within a middleware is a common practice, and once we refactor mojito middleware infrastructure, in which case we know we will break BC, we can remove the sugar.

caridy commented on lib/app/middleware/mojito-handler-tunnel-type.js:
```diff
@@ -0,0 +1,56 @@
+module.exports = function (config) {
+    return function (req, res, next) {
+        var typeReq  = req._tunnel && req._tunnel.typeReq,
+            instance = {};
```
same as above.

caridy commented on lib/app/middleware/mojito-handler-tunnel-type.js:
```diff
@@ -0,0 +1,56 @@
+        }
+
+        if (!typeReq.type) {
+            return sendError(res, 'Not found: ' + req.url, 500);
```
same as above

caridy commented on lib/app/middleware/mojito-handler-tunnel-rpc.js:
```diff
@@ -0,0 +1,66 @@
+        var command  = req.body,
+            instance = command.instance;
+
+        command.context = command.context || {};
```
move logic after the `if`

caridy commented:
done with the first pass @ekashida

ekashida committed:
implement feedback from pull #1044

- delay object creation for performance reasons
- use core library methods for parsing url/path
- fix http response code (500 => 404)
- eliminate `that = this` usage by closing over prefix variables directly

ekashida committed:
clean up the logic around path inspection

- Replaced the brittle `trimSlash` stuff with `require('path').normalize`.

ekashida committed:
replace tunnel middleware with refactored version

- Remove trailing slashes from tunnel/static prefixes.
- Add flag and flag check to determine when to stop processing the request.

ekashida commented:
@caridy I think I've addressed all your feedback, although I'm still getting those weird unit test errors. I'll keep looking at them...

ekashida committed:
fix bug where configured prefixes can be undefined

ekashida committed:
fix test mock and remove unused variables

ekashida committed:
execute tunnel middleware as an async queue

caridy commented on lib/app/middleware/mojito-handler-tunnel-rpc.js:
```diff
@@ -0,0 +1,58 @@
+            command.instance,
+            command.context,
+            function (err, instance) {
+                // Replace with the expanded instance.
```
If `err` is detected, we should call next(err) at least.

caridy commented on lib/app/middleware/mojito-handler-tunnel-parser.js:
```diff
@@ -0,0 +1,108 @@
+
+            // "Spec" tunnel request
+            if (parts[parts.length - 2] === 'specs') {
+                req._tunnel.specsReq = {
```
this is a bold assumption, what is req._tunnel is not set? Why is this middleware coupled with others?

caridy commented on lib/app/middleware/mojito-handler-tunnel-specs.js:
```diff
@@ -0,0 +1,58 @@
+        name = specsReq.name;
+
+        if (!type || !name) {
+            return req._tunnel.sendError(res, 'Not found: ' + req.url, 404);
```
same as above, bold assumption. I think we should call `next(new Error())` instead.

caridy commented on lib/app/middleware/mojito-handler-tunnel-type.js:
```diff
@@ -0,0 +1,49 @@
+            req.context,
+            function (err, data) {
+                if (err) {
+                    return req._tunnel.sendError(
```
same as above.

caridy commented on lib/app/middleware/mojito-handler-tunnel-type.js:
```diff
@@ -0,0 +1,49 @@
+        }
+
+        if (!typeReq.type) {
+            return req._tunnel.sendError(res, 'Not found: ' + req.url, 404);
```
same as above.

caridy commented on lib/app/middleware/mojito-handler-tunnel.js:
```diff
@@ -4,202 +4,57 @@
+        ];
+
+        // Helper methods.
+        req._tunnel = {
```
If this is not used in this middleware, it should not be defined here.

ekashida committed:
streamline middleware error handling

Uses the convention where a middleware with an arity of 4 is assumed to be the
error handler.

ekashida committed:
initial unit test commit for refactored tunnel middleware

ekashida committed:
simplify response logic with express response sugar

With unit tests!
